File: src/minification.c
/* Copyright 2007 Walter Alini, Matías Bordese */

/*
This file is part of biOps.

    biOps is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    biOps is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with biOps; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/


/*
	Title: Minification Operations (C code)
*/

#include <stdlib.h>
#include "defs.h"

#define ELEM_SWAP(a,b) { register int t=(a);(a)=(b);(b)=t; }

int quick_select(int *arr, int n){
	int low = 0, high = n - 1;
	int median = (low + high) / 2;
	int middle, ll, hh;

	while (1){
		if (high <= low)
			return arr[median]; /* One element only */
		if (high == low + 1){
			if (arr[low] > arr[high])
				ELEM_SWAP(arr[low], arr[high]);
			return arr[median];
		}

		/* Find median of low, middle and high items; swap into position low */
		middle = (low + high) / 2;
		if (arr[middle] > arr[high])
			ELEM_SWAP(arr[middle], arr[high]) ;
		if (arr[low] > arr[high])
			ELEM_SWAP(arr[low], arr[high]) ;
		if (arr[middle] > arr[low])
			ELEM_SWAP(arr[middle], arr[low]) ;

		/* Swap low item (now in position middle) into position (low+1) */
		ELEM_SWAP(arr[middle], arr[low+1]) ;

		/* Nibble from each end towards middle, swapping items when stuck */
		ll = low + 1;
		hh = high;
		while (1){
			do ll++;
			while (arr[low] > arr[ll]);
			do hh--;
			while (arr[hh]  > arr[low]);

			if (hh < ll)
				break;

			ELEM_SWAP(arr[ll], arr[hh]) ;
		}

		/* Swap middle item (in position low) back into correct position */
		ELEM_SWAP(arr[low], arr[hh]) ;

		/* Re-set active partition */
		if (hh <= median)
			low = ll;
		if (hh >= median)
			high = hh - 1;
	}
}


double window_average(int arr[], int n){
	int i;
	double ret = 0;

	for(i = 0; i < n; i++){
		ret = ret + arr[i];
	}

	return (ret/n);
}


/*
	Function: medianShrink
	Performs a shrink operation using median representation

	Parameters:
		image - The input image
		width - The image width
		height - Its height
		depth - Its color depth
		ret - The result reduced image
		new_width - The width of the reduced image
		new_height - The height

	Returns:
		The reduced image
*/
void medianShrink(int *image, int *width, int *height, int *depth, int *ret, int *new_width, int *new_height){
	int new_rows = *new_height, new_cols = *new_width;
	int x_side = *width / new_cols, y_side = (*height / new_rows), wsize = x_side * y_side;
	int *window = calloc(wsize, sizeof(double));
	int x, y, d, winx, winy;
	double value;

	for (y = 0; y < new_rows; y++){
		for (x = 0; x < new_cols; x++){
			for (d = 0; d < *depth; d++){
				int windex = 0;
				for (winy = 0; winy < y_side; winy++){
					for (winx = 0; winx < x_side; winx++){
						window[windex] = image[IMGPOS(x * x_side + winx, y * y_side + winy, d, *width, *height)];
						windex++;
					}
				}
				value = quick_select(window, windex);
				ret[IMGPOS(x, y, d, new_cols, new_rows)] = (int) value;
			}
		}
	}

	free(window);
}

/*
	Function: averageShrink
	Performs a shrink operation using average representation

	Parameters:
		image - The input image
		width - The image width
		height - Its height
		depth - Its color depth
		ret - The result reduced image
		new_width - The width of the reduced image
		new_height - The height

	Returns:
		The reduced image
*/
void averageShrink(int *image, int *width, int *height, int *depth, int *ret, int *new_width, int *new_height){
	int new_rows = *new_height, new_cols = *new_width;
	int x_side = *width / new_cols, y_side = (*height / new_rows), wsize = x_side * y_side;
	int *window = calloc(wsize, sizeof(double));
	int x, y, d, winx, winy;
	double value;

	for (y = 0; y < new_rows; y++){
		for (x = 0; x < new_cols; x++){
			for (d = 0; d < *depth; d++){
				int windex = 0;
				for (winy = 0; winy < y_side; winy++){
					for (winx = 0; winx < x_side; winx++){
						window[windex] = image[IMGPOS(x * x_side + winx, y * y_side + winy, d, *width, *height)];
						windex++;
					}
				}
				value = window_average(window, windex);
				ret[IMGPOS(x, y, d, new_cols, new_rows)] = (int) value;
			}
		}
	}

	free(window);
}

File: R/minification.R
# Copyright 2007 Walter Alini, Matías Bordese

#
# This file is part of biOps.
#
#     biOps is free software; you can redistribute it and/or modify
#     it under the terms of the GNU General Public License as published by
#     the Free Software Foundation; either version 2 of the License, or
#     (at your option) any later version.
#
#     biOps is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with biOps; if not, write to the Free Software
#     Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#


#
#	Title: Minification Operations
#

.check_shrink_factors <- function(x_shrink, y_shrink){
	if (x_shrink <= 0 || y_shrink <= 0)
		stop("Shrink factors must be positive")
	if(x_shrink > 1 || y_shrink > 1)
		stop("Shrink factors must be less than 1")
}

#	Function: imgMedianShrink
#	Reduce an image using median
#
#	Parameters:
#		imgdata - The image data.
#		x_scale - Horizontal axis scale factor.
#		y_scale - Vertical axis scale factor.
#
#	Returns:
#		The reduced image.
#
imgMedianShrink <- function(imgdata, x_scale, y_scale){
	.imgShrink(imgdata, x_scale, y_scale, 'median')
}

#	Function: imgAverageShrink
#	Reduce an image using average
#
#	Parameters:
#		imgdata - The image data.
#		x_scale - Horizontal axis scale factor.
#		y_scale - Vertical axis scale factor.
#
#	Returns:
#		The reduced image.
#
imgAverageShrink <- function(imgdata, x_scale, y_scale){
	.imgShrink(imgdata, x_scale, y_scale, 'average')
}

#	Function: imgShrink
#	Reduce an image
#
#	Parameters:
#		imgdata - The image data.
#		x_scale - Horizontal axis scale factor.
#		y_scale - Vertical axis scale factor.
#		reduction - The method for reduction: average | median.
#
#	Returns:
#		The reduced image.
#
.imgShrink <- function(imgdata, x_scale, y_scale, reduction='average'){
	method <- switch(reduction, average='averageShrink', median='medianShrink')
	if (is.null(method)) stop('Unsupported reduction method')
	.check_shrink_factors(x_scale, y_scale)
	imgmatrix <- array(imgdata) # get linear array image representations
	depth <- if (attr(imgdata, "type") == "grey") 1 else dim(imgdata)[3] # get images depth
	width <- dim(imgdata)[2]
	height <- dim(imgdata)[1]

	# dims of the result
	cols <- as.integer(width * x_scale)
	rows <- as.integer(height * y_scale)

	# call the C function for image operation
	res <- .C(method, image=as.integer(imgmatrix),
			width=as.integer(width), height=as.integer(height), depth=as.integer(depth),
			ret=integer(cols * rows * depth), new_width=as.integer(cols), new_height=as.integer(rows),
			PACKAGE="biOps")

	imgtype <- if (depth == 1) "grey" else "rgb" # type of the result
	imgdim <- c(rows, cols, if (depth == 3) depth else NULL) # dim of the result
	img <- array(res$ret, dim=imgdim) # build the matrix from linear result
	imagedata(img, type=imgtype) # build the imagedata
}
